「日報の言及機能を作る」の課題を作成しました

a-kuroki-gs opened the pull request.

a-kuroki-gs committed:
言及先、言及元の特定に必要なモデルを作成

a-kuroki-gs committed:
モデルの中身を一部変更

mentioning_report.rbとmentioned_report.rbから使うテーブルをreportにするため
削除の関連付けができるようにするため

a-kuroki-gs committed:
不要なモデルを削除し、正しく関連付けを行うよう修正

a-kuroki-gs committed:
Report削除→Mentionも削除、となるように修正

a-kuroki-gs committed:
日報の作成・編集時に、日報のリンクを読み込んで言及機能として表示できるようにする

a-kuroki-gs committed:
DBのユニーク制約やモデルのバリデーションを検討

a-kuroki-gs committed:
rubocop・ERBLintの対応を行う

yuuu commented on app/views/reports/show.html.erb:
```diff
@@ -4,6 +4,15 @@
 </div>
 
+<% if Mention.where(mentioning_report_id: @report.id).exists? %>
+  <h2>この日報は以下の日報に言及されています</h2>
```
こちらはi18n対応が漏れていますね 👀

a-kuroki-gs replied:
漏れていたi18n対応を行いました。

yuuu commented on db/migrate/20231113054023_add_index_to_mentions.rb:
```diff
@@ -0,0 +1,5 @@
+class AddIndexToMentions < ActiveRecord::Migration[7.0]
+  def change
+    add_index :mentions, [:mentioned_report_id, :mentioning_report_id], unique: true
```
複合Indexが貼られていて良いですね 👍

yuuu commented on app/models/report.rb:
```diff
@@ -4,6 +4,12 @@ class Report < ApplicationRecord
+  has_many :mentioned_reports, through: :active_relationship
+
+  has_many :passive_relationship, class_name: 'Mention', foreign_key: :mentioned_report_id, dependent: :destroy, inverse_of: :mentioned_report
+  has_many :mentioning_reports, through: :passive_relationship
```
`passive` は「受け身」という意味なので、こちらが `mentioned_reports` ではないでしょうか？

a-kuroki-gs replied:
あまり意味を考えずに使ってしまっていました。

active：能動的、passive：受動的ということで、
`active_relationship`を`mentioning_reports`をに、`passive_relationship`を`mentioned_reports`に変更しました。

yuuu commented on app/views/reports/show.html.erb:
```diff
@@ -4,6 +4,15 @@
   <%= render @report %>
 </div>
 
+<% if Mention.where(mentioning_report_id: @report.id).exists? %>
```
ここは `@report.mentioned_reports.exists?` のように書くと読みやすいですね 👍

a-kuroki-gs replied:
`mention`そのものを見つけないと…という意識しかなく、
`report.mentionied_reports`で探すのは考えていませんでしたので、勉強になりました。
修正しております。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -30,6 +39,18 @@ def create
+      mentions = Mention.where(mentioned_report_id: @report.id)
+      mentions.each(&:destroy)
+
+      contents = @report.content.scan(%r{http://localhost:3000/reports/(\d{1,})})
```
ここは `contents` というより `report_ids` といった名前のほうが適切ですね。

yuuu replied:
結果が二次元配列になってしまっているので、この段階で `flatten` しておくと、あとで `[0]` と書く必要がなくなりますね。

もしくは正規表現の後読みを使うという手もあります。こちらは参考まで。

```ruby
%r{(?<=http://localhost:3000/reports/)\d+}
```

yuuu replied:
`create` `update` はメンションの作成・更新に対応していますが、 `destroy` 時はメンションを操作する必要はないでしょうか？

a-kuroki-gs replied:
変数名を`report_ids`に修正し、
二次元配列にならないように`flatten`を利用しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -30,6 +39,18 @@ def create
   def update
     if @report.update(report_params)
+      mentions = Mention.where(mentioned_report_id: @report.id)
+      mentions.each(&:destroy)
```
`destroy_all` というメソッドがあるので、そちらを使うとスッキリ書けますね 👍

a-kuroki-gs replied:
`destroy_all`メソッドを利用し修正しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -30,6 +39,18 @@ def create
 
   def update
     if @report.update(report_params)
+      mentions = Mention.where(mentioned_report_id: @report.id)
```
ここも `@report.mentioning_report` とすると何を取得しようとしているかがわかりやすいです。

a-kuroki-gs replied:
先ほどと同様に`mention`自体を削除しようとしていたので、
`@report.mentioning_report`を削除するよう修正しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -22,6 +22,15 @@ def create
 
     if @report.save
+      contents = @report.content.scan(%r{http://localhost:3000/reports/(\d{1,})})
+      unless contents.empty?
```
`empty?` な場合は `contents.each` のブロックが1回も実行されないだけなので、このチェックは省略できるはずです。

a-kuroki-gs replied:
不要でしたので省略しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -22,6 +22,15 @@ def create
     @report = current_user.reports.new(report_params)
 
     if @report.save
+      contents = @report.content.scan(%r{http://localhost:3000/reports/(\d{1,})})
```
ここからメンションを作成するまでの処理は、1つの日報に対して行う処理なので `Report` クラスに持っていきたいですね。 `update` 時の処理についても同様です。

a-kuroki-gs replied:
`Report`クラスで`save_mention`メソッドを定義しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -22,6 +22,15 @@ def create
+          mention = Mention.new
+          mention.mentioning_report_id = content[0]
+          mention.mentioned_report_id = @report.id
+          mention.save
```
今の実装だと、「必須要件」に書かれている

> 日報の保存には成功したが、言及データの保存には失敗している、という状況が発生しないようにする

を防げていないはずなので、これを防ぐよう対応をお願いします 🙏

a-kuroki-gs replied:
トランザクションを実装し、必須要件を満たすように修正しました。

a-kuroki-gs committed:
updateメソッドのわかりにくい部分を修正

すべての要素を削除するのにdestroy_allメソッドを利用する
日報の言及先をreport.mentioning_reportsで取得し、それを削除する

a-kuroki-gs committed:
日報のリンクを取得していた部分を修正

Reportクラスでまとめて定義
不要なチェックの削除
flattenを用いて二次元配列にならないようにする
正規表現で取得した日報のIDを入れる配列名をreport_idsとしてわかりやすくする
createメソッド、updateメソッドにsave_mentionを適用

a-kuroki-gs committed:
「日報の保存には成功したが、言及データの保存には失敗している、という状況が発生しないようにする」をカバーできるよう修正

a-kuroki-gs committed:
ビューにおいてMentionから言及元を探すのではなく、report.mentioned_reportsで探すようにする

a-kuroki-gs committed:
モデルの関連付けにおいて、activeとpassiveを正しく定義する

a-kuroki-gs committed:
i18n対応を行う

a-kuroki-gs committed:
言及元の日報をID順に並べて表示する

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -21,19 +21,27 @@ def edit; end
-      render :new, status: :unprocessable_entity
+    ActiveRecord::Base.transaction do
+      @report.save
+      @report.save_mention
```
ここでの保存に失敗したときに `@report.save` もなかったことにしたい、というのがトランザクションを用いた目的だと思いますが、今の実装だと `@report.save_mention` が例外を出さないので目的が達成されていない( `@report.save` が巻き戻らない)と思われます。

`@report.save_mention` で例外を出すよう修正しましょう。`update` 時も同様です。

a-kuroki-gs replied:
`save_mention`メソッド内で、保存に失敗した場合に例外を発生させる処理を追加しました。
まだ更新時には、メンション情報を削除した際にも例外を発生させるよう、
`raise ActiveRecord::RecordNotSaved unless @report.mentioning_reports.destroy_all`と変更しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -21,19 +21,27 @@ def edit; end
+      @report.save_mention
     end
+    redirect_to @report, notice: t('controllers.common.notice_create', name: Report.model_name.human)
+  rescue StandardError
```
`ActiveRecord::Base.transaction` で発生しうる例外は限られるはずなので、 `rescue` する例外をもう少し絞り込めないか検討ください。 `update` 時も同様です。

a-kuroki-gs replied:
https://github.com/a-kuroki-gs/fjord-books_app-2023/pull/6#discussion_r1408653384
こちらで発生させた例外(`RecordNotSaved`)のみ記載しました。

yuuu commented on app/models/report.rb:
```diff
@@ -14,4 +20,14 @@ def editable?(target_user)
+    report_ids = content.scan(%r{http://localhost:3000/reports/(\d{1,})}).flatten
+    report_ids.each do |report_id|
+      mention = Mention.new
+      mention.mentioning_report_id = report_id
```
このように生成すると短く書けますし、何をしているのかわかりやすくなりますね。

```ruby
self.active_relationship.create(mentioning_report_id: report_id)
```

別解としてこんな書き方もあります。

```ruby
reports = Report.where(id: report_ids)
self.mentioning_reports += reports
```

yuuu replied:
今の実装だと、この日報自身のURLが含まれている場合にも言及が作成されてしまうので、この日報だけは除外する処理を実装する必要がありますね。

a-kuroki-gs replied:
> このように生成すると短く書けますし、何をしているのかわかりやすくなりますね。

> 別解としてこんな書き方もあります。

`Mention`を作成して紐づけて…としていたため、冗長なメソッドになっておりました。
個人的に後者のコードが好きだったので、こちらに修正しました。

> 今の実装だと、この日報自身のURLが含まれている場合にも言及が作成されてしまうので、この日報だけは除外する処理を実装する必要がありますね。

こちら対応できておりませんでしたので、その日報自身は省くように修正しました。

a-kuroki-gs committed:
メンション情報を保存するメソッドの内部を端的にまとめる

mentionを新しく作成するのではなく、reportの属性を追加するように修正

a-kuroki-gs committed:
その日報自体を言及できないようにする

a-kuroki-gs committed:
言及の保存に失敗したときに、例外を発生させる処理を追加

save_mentionメソッドに追加し、更新時はメンションの削除処理失敗時に例外を発生させるよう修正
rescueする例外も発生させた例外のみキャッチする

a-kuroki-gs committed:
rubocop対応

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -26,20 +26,20 @@ def create
   def update
     ActiveRecord::Base.transaction do
       @report.update(report_params)
+      raise ActiveRecord::RecordNotSaved unless @report.mentioning_reports.destroy_all
```
こういうときにraiseする例外は `ActiveRecord::Rollback` を使うと良いです

a-kuroki-gs replied:
トランザクション周り、修正しました。
詳細は[こちら](https://bootcamp.fjord.jp/products/18026)に記載しております。

a-kuroki-gs committed:
ActiveRecord::Base.transaction do 〜 endの処理をモデルクラス内でメソッドとして定義

a-kuroki-gs committed:
rescue節を消去し、if文で保存できたか判定するように修正

a-kuroki-gs committed:
save_mentionメソッドの発生させる例外をActiveRecord::Rollbackに変更

rescue節を消去したことにより、ActiveRecord::RecordNotSavedの例外は拾えなくなったため
その後のif文でtrue / falseの値が必要なため

yuuu commented on app/models/report.rb:
```diff
@@ -24,6 +24,24 @@ def created_on
     report_ids = content.scan(%r{http://localhost:3000/reports/(\d{1,})}).flatten
     reports = Report.where.not(id:).where(id: report_ids)
-    raise ActiveRecord::RecordNotSaved unless self.mentioning_reports += reports
+    raise ActiveRecord::Rollback unless self.mentioning_reports += reports
```
ここですが、おそらくレコードの作成に失敗したときは `+=` メソッドの内部で `ActiveRecord::Rollback` が `raise` されるものと思われます。

https://github.com/rails/rails/blob/b3b230c08e5f75516561c727ed2a2735319ab60a/activerecord/lib/active_record/associations/collection_association.rb#L444

ソースコードを追い間違えている可能性もあるので、意図的に追加に失敗させて(コードを一時的に変更し、あえて同じ日報を追加するなど)どういった挙動となるか確認いただけると助かります 🙏 

内部で `raise` されるのであれば、この行の `raise` や `unless` は不要となるはずです。

a-kuroki-gs replied:
`ActiveRecord::RecordInvalid`が`raise`されていたので、`unless`文を消去しました。

その他、[こちら](https://bootcamp.fjord.jp/products/18026)に記載しております。

yuuu commented on app/models/report.rb:
```diff
@@ -24,6 +24,24 @@ def created_on
+    raise ActiveRecord::Rollback unless self.mentioning_reports += reports
+  end
+
+  def create_trunsaction
```
`trunsaction` がtypoしていますね。正しくは `transaction` かと 👍 
update側も併せて修正ください。

a-kuroki-gs committed:
transactionのtypoを修正

a-kuroki-gs committed:
不要なunlessやraiseを消去

+=に失敗した場合は内部でActiveRecord::RecordInvalidをraiseしてくれるため

a-kuroki-gs committed:
rescue節をコントローラ側に記述

true / falseでの処理を消去したため
フラッシュメッセージやリダイレクト等の処理はコントローラ側で行うため

a-kuroki-gs committed:
saveの例外を考慮する

念のため

a-kuroki-gs committed:
レコードの作成に失敗したときの処理を修正

処理自体から例外が発生する場合はraise等不要なので消去
rescue節でtrue / falseを返すように修正し、その結果に応じてコントローラ側で処理を行う

yuuu commented on app/models/report.rb:
```diff
@@ -24,24 +24,28 @@ def created_on
+  def update_transaction(report_params)
     ActiveRecord::Base.transaction do
       update(report_params)
-      raise ActiveRecord::Rollback unless mentioning_reports.destroy_all
```
ここが `unless` → `if` に変わっている点が気になって調べたのですが、 `destroy_all` は削除できたかどうかを戻り値として返してくれないようですね。

この件も含めて、以下記事で伊藤さんがまとめてくださっているので、こちらを読んで修正をお願い致します。
https://qiita.com/jnchito/items/3393c5c1a744199e128a#%E5%8F%82%E8%80%832-destroy_all%E3%82%92%E4%BD%BF%E3%81%86%E5%A0%B4%E5%90%88%E3%81%AE%E6%B3%A8%E6%84%8F%E7%82%B9

a-kuroki-gs replied:
すみません、こちらテスト用に一時的に`if`にしたのをコミットしてしまいました。
また、`destroy_all`の返り値についても把握できておりませんでした。

削除操作→削除できたか確認→できていなかったら例外を発生させる、という処理に変更しました。

a-kuroki-gs committed:
destroy_all周りを修正

テスト用に変更したifがそのままになっていたため
destroy_allの返り値は真偽値ではないため
